File: src/main/java/com/example/inventory/api/ProductController.java
package com.example.inventory.api;

import com.example.inventory.command.AddInventoryCommand;
import com.example.inventory.command.CreateProductCommand;
import com.example.inventory.command.RemoveInventoryCommand;
import com.example.inventory.query.FindAllProductsQuery;
import com.example.inventory.query.FindProductByIdQuery;
import com.example.inventory.query.ProductQueryEntity;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.axonframework.queryhandling.QueryGateway;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

@RestController
@RequestMapping("/api/products")
public class ProductController {

    private final CommandGateway commandGateway;
    private final QueryGateway queryGateway;

    public ProductController(CommandGateway commandGateway, QueryGateway queryGateway) {
        this.commandGateway = commandGateway;
        this.queryGateway = queryGateway;
    }

    // 商品作成
    @PostMapping
    public CompletableFuture<String> createProduct(@RequestBody CreateProductRequest request) {
        String productId = UUID.randomUUID().toString();
        return commandGateway.send(new CreateProductCommand(
                productId,
                request.getName(),
                request.getInitialQuantity()
        ));
    }

    // 在庫追加
    @PostMapping("/{productId}/add-inventory")
    public CompletableFuture<Void> addInventory(
            @PathVariable String productId,
            @RequestBody UpdateInventoryRequest request) {
        return commandGateway.send(new AddInventoryCommand(productId, request.getQuantity()));
    }

    // 在庫減少
    @PostMapping("/{productId}/remove-inventory")
    public CompletableFuture<Void> removeInventory(
            @PathVariable String productId,
            @RequestBody UpdateInventoryRequest request) {
        return commandGateway.send(new RemoveInventoryCommand(productId, request.getQuantity()));
    }

    // すべての商品を取得
    @GetMapping
    public CompletableFuture<List<ProductQueryEntity>> getAllProducts() {
        return queryGateway.query(
                new FindAllProductsQuery(),
                org.axonframework.messaging.responsetypes.ResponseTypes.multipleInstancesOf(ProductQueryEntity.class)
        );
    }

    // 特定の商品を取得
    @GetMapping("/{productId}")
    public CompletableFuture<ResponseEntity<ProductQueryEntity>> getProduct(@PathVariable String productId) {
        return queryGateway.query(
                new FindProductByIdQuery(productId),
                org.axonframework.messaging.responsetypes.ResponseTypes.optionalInstanceOf(ProductQueryEntity.class)
        ).thenApply(optionalProduct ->
                optionalProduct.map(ResponseEntity::ok)
                        .orElse(ResponseEntity.notFound().build())
        );
    }
}

File: src/main/kotlin/com/example/inventory/api/ProductController.kt
package com.example.inventory.api

import com.example.inventory.command.AddInventoryCommand
import com.example.inventory.command.CreateProductCommand
import com.example.inventory.command.RemoveInventoryCommand
import com.example.inventory.query.FindAllProductsQuery
import com.example.inventory.query.FindProductByIdQuery
import com.example.inventory.query.ProductQueryEntity
import org.axonframework.commandhandling.gateway.CommandGateway
import org.axonframework.messaging.responsetypes.ResponseTypes
import org.axonframework.queryhandling.QueryGateway
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*
import java.util.concurrent.CompletableFuture

@RestController
@RequestMapping("/api/products")
class ProductController(
    private val commandGateway: CommandGateway,
    private val queryGateway: QueryGateway
) {

    // 商品作成
    @PostMapping
    fun createProduct(@RequestBody request: CreateProductRequest): CompletableFuture<String> {
        val productId = UUID.randomUUID().toString()
        return commandGateway.send(
            CreateProductCommand(
                productId,
                request.name,
                request.initialQuantity
            )
        )
    }

    // 在庫追加
    @PostMapping("/{productId}/add-inventory")
    fun addInventory(
        @PathVariable productId: String,
        @RequestBody request: UpdateInventoryRequest
    ): CompletableFuture<Void> {
        return commandGateway.send(AddInventoryCommand(productId, request.quantity))
    }

    // 在庫減少
    @PostMapping("/{productId}/remove-inventory")
    fun removeInventory(
        @PathVariable productId: String,
        @RequestBody request: UpdateInventoryRequest
    ): CompletableFuture<Void> {
        return commandGateway.send(RemoveInventoryCommand(productId, request.quantity))
    }

    // すべての商品を取得
    @GetMapping
    fun getAllProducts(): CompletableFuture<List<ProductQueryEntity>> {
        return queryGateway.query(
            FindAllProductsQuery(),
            ResponseTypes.multipleInstancesOf(ProductQueryEntity::class.java)
        )
    }

    // 特定の商品を取得
    @GetMapping("/{productId}")
    fun getProduct(@PathVariable productId: String): CompletableFuture<ResponseEntity<ProductQueryEntity>> {
        return queryGateway.query(
            FindProductByIdQuery(productId),
            ResponseTypes.optionalInstanceOf(ProductQueryEntity::class.java)
        ).thenApply { optionalProduct ->
            optionalProduct.map { ResponseEntity.ok(it) }
                .orElse(ResponseEntity.notFound().build())
        }
    }
}
